src/chat/chatReducers.js

```diff
@@ -70,6 +72,29 @@ const eventMessageDelete = (
   return stateChange ? newState : state;
 };
 
+const eventUpdateMessageFlags = (
+  state: NarrowsState,
+  action: EventUpdateMessageFlagsAction,
+): NarrowsState => {
+  const { messages, flag, operation } = action;
+  const sources = [state];
+
+  if (state[STARRED_NARROW_STR] && flag === 'starred') {
+    if (operation === 'add') {
+      sources.push({
+        [STARRED_NARROW_STR]: [...state[STARRED_NARROW_STR], ...messages].sort(),
+      });
+    } else {
+      // operation === 'remove'
+      sources.push({
+        [STARRED_NARROW_STR]: state[STARRED_NARROW_STR].filter(id => !messages.includes(id)),
```

Comment:
Probably rarely a big deal in practice: but this is quadratic, in the number of already-known starred messages and the number we're learning were unstarred.

How about `const messagesSet = new Set(messages);` just above, and then use `messagesSet`?

Reply:
Updated.